File: plugin/tongs-runner/src/main/java/com/github/tarcv/tongs/Tongs.java
/*
 * Copyright 2020 TarCV
 * Copyright 2014 Shazam Entertainment Limited
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
 * the License.
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */
package com.github.tarcv.tongs;

import com.github.tarcv.tongs.injector.RuleHelpersKt;
import com.github.tarcv.tongs.injector.TongsRunnerInjector;
import com.github.tarcv.tongs.runner.rules.RunRule;
import com.github.tarcv.tongs.runner.rules.RunRuleContext;
import com.github.tarcv.tongs.runner.rules.RunRuleFactory;
import com.google.common.collect.Lists;
import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.util.List;

import static com.github.tarcv.tongs.injector.ConfigurationInjector.configuration;
import static com.github.tarcv.tongs.injector.ConfigurationInjector.setConfiguration;
import static com.github.tarcv.tongs.injector.RuleManagerFactory.fixGenericClass;
import static com.github.tarcv.tongs.injector.TongsRunnerInjector.tongsRunner;
import static com.github.tarcv.tongs.utils.Utils.millisSinceNanoTime;
import static java.lang.System.nanoTime;
import static org.apache.commons.io.FileUtils.deleteDirectory;
import static org.apache.commons.lang3.time.DurationFormatUtils.formatPeriod;

public final class Tongs {
    private static final Logger logger = LoggerFactory.getLogger(Tongs.class);

    private final TongsRunner tongsRunner;
    private final File output;

    public Tongs(Configuration configuration) {
        this.output = configuration.getOutput();
        setConfiguration(configuration);
        this.tongsRunner = TongsRunnerInjector.tongsRunner();
    }

    public boolean run() {
		long startOfTestsMs = nanoTime();
        List<RunRule> runRules = RuleHelpersKt.ruleManagerFactory()
                .<RunRuleContext, RunRule, RunRuleFactory<RunRule>>create(
                        fixGenericClass(RunRuleFactory.class),
                        (runRuleFactory, runRuleContext) -> {
                            return runRuleFactory.runRules(runRuleContext);
                        }
                )
                .createRulesFrom((configuration) -> new RunRuleContext(configuration));
        runRules.forEach(runRule -> runRule.before());
		try {
            deleteDirectory(output);
            //noinspection ResultOfMethodCallIgnored
            output.mkdirs();
            return tongsRunner.run();
		} catch (Exception e) {
            logger.error("Error while running Tongs", e);
			return false;
		} finally {
            long duration = millisSinceNanoTime(startOfTestsMs);
            logger.info(formatPeriod(0, duration, "'Total time taken:' H 'hours' m 'minutes' s 'seconds'"));
            Lists.reverse(runRules)
                    .forEach(runRule -> runRule.after());
        }
	}
}

File: plugin/tongs-runner/src/main/java/com/github/tarcv/tongs/Tongs.kt
/*
 * Copyright 2020 TarCV
 * Copyright 2014 Shazam Entertainment Limited
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
 * the License.
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */
package com.github.tarcv.tongs

import com.github.tarcv.tongs.injector.RuleManagerFactory
import com.github.tarcv.tongs.injector.TongsRunnerInjector.createTongsRunner
import com.github.tarcv.tongs.injector.withRules
import com.github.tarcv.tongs.runner.AndroidDdmRunRuleFactory
import com.github.tarcv.tongs.api.run.RunConfiguration
import com.github.tarcv.tongs.api.run.RunRule
import com.github.tarcv.tongs.api.run.RunRuleContext
import com.github.tarcv.tongs.api.run.RunRuleFactory
import org.apache.commons.io.FileUtils
import org.apache.commons.lang3.time.DurationFormatUtils
import org.koin.core.context.KoinContextHandler
import org.koin.core.context.startKoin
import org.koin.core.context.stopKoin
import org.koin.dsl.module
import org.slf4j.LoggerFactory
import java.io.File

class Tongs(configuration: Configuration) {
    private val runnerModule = module {
        single { configuration }
        single {
            val conf = get<Configuration>()
            RuleManagerFactory(conf, conf.pluginsInstances)
        }
        single { createTongsRunner(get()) }
    }

    @JvmOverloads
    fun run(allowThrows: Boolean = false): Boolean {
        startKoin {
            modules(runnerModule)
        }
        val tongsRunner by KoinContextHandler.get().inject<TongsRunner>()
        try {
            val startOfTestsMs = System.nanoTime()
            val predefinedRulesFactories = listOf(
                    PrepareOutputDirectoryRuleFactory(),
                    AndroidDdmRunRuleFactory()
            )
            val ruleManagerFactory by KoinContextHandler.get().inject<RuleManagerFactory>()
            val runRules = ruleManagerFactory
                    .create<RunRuleContext, RunRule, RunRuleFactory<RunRule>>(
                            RunRuleFactory::class.java,
                            predefinedRulesFactories
                    ) { runRuleFactory: RunRuleFactory<RunRule>, runRuleContext: RunRuleContext ->
                        runRuleFactory.runRules(runRuleContext)
                    }
                    .createRulesFrom { configuration: RunConfiguration ->
                        RunRuleContext(configuration)
                    }

            return try {
                withRules(
                        logger,
                        "while executing a run rule",
                        "while running Tongs",
                        runRules,
                        { it.before() },
                        { it, ret ->
                            it.after()
                            ret
                        }
                ) {
                    if (allowThrows) {
                        tongsRunner.throwingRun()
                    } else {
                        tongsRunner.run()
                    }
                }
            } catch (e: Exception) {
                if (allowThrows) {
                    throw e
                } else {
                    false
                }
            } finally {
                val duration = Utils.millisSinceNanoTime(startOfTestsMs)
                logger.info(DurationFormatUtils.formatPeriod(0, duration,
                        "'Total time taken:' H 'hours' m 'minutes' s 'seconds'"))
            }
        } finally {
            stopKoin()
        }
    }

    class PrepareOutputDirectoryRuleFactory(): RunRuleFactory<PrepareOutputDirectoryRule> {
        override fun runRules(context: RunRuleContext): Array<out PrepareOutputDirectoryRule> {
            return arrayOf(PrepareOutputDirectoryRule(context.configuration.output))
        }
    }

    class PrepareOutputDirectoryRule(private val output: File): RunRule {
        override fun before() {
            FileUtils.deleteDirectory(output)
            //noinspection ResultOfMethodCallIgnored
            output.mkdirs()
        }

        override fun after() {
            // no-op
        }
    }

    companion object {
        private val logger = LoggerFactory.getLogger(Tongs::class.java)
    }
}